File: web/src/domain/resolver.ts
// Deterministic Resolver — semantic Moment → semantic Trigger, and a semantic
// Trigger → concrete per-place Triggers (1 label → N places).
// No AI here.

import { isSimulatable } from './places';
import { placeKey } from './placeDictionary';
import type { MomentCandidate, PlaceDict, PlaceId, PlaceRef, Trigger } from './types';

export type ResolveResult =
  | { ok: true; trigger: Trigger }
  | { ok: false; reason: string; needsLearning?: boolean; phrase?: string };

/** 候補 → semantic Trigger（場所は意味ラベル or アンカーのまま）。 */
export function resolve(c: MomentCandidate, dict: PlaceDict): ResolveResult {
  switch (c.kind) {
    case 'time':
      if (!c.timeBucket) return { ok: false, reason: '時間帯が特定できませんでした' };
      return { ok: true, trigger: { primitive: 'time', timeBucket: c.timeBucket } };

    case 'home_arrival':
      return anchorTrigger('place_enter', 'home');
    case 'leave_home':
      return anchorTrigger('place_exit', 'home');
    case 'work_arrival':
      return anchorTrigger('place_enter', 'work');

    case 'place_arrival':
    case 'place_departure': {
      const primitive = c.kind === 'place_departure' ? 'place_exit' : 'place_enter';

      if (c.anchorHint) return anchorTrigger(primitive, c.anchorHint);

      const phrase = c.placePhrase ?? c.placeLabel;
      if (!phrase) return { ok: false, reason: '場所が特定できませんでした' };

      const key = placeKey(phrase);
      if ((dict[key]?.length ?? 0) > 0) {
        return { ok: true, trigger: { primitive, ref: { kind: 'label', key } } };
      }
      return {
        ok: false,
        reason: `「${phrase}」がどこか、まだ分かりません`,
        needsLearning: true,
        phrase,
      };
    }
  }
}

function anchorTrigger(
  primitive: 'place_enter' | 'place_exit',
  anchor: 'home' | 'work',
): ResolveResult {
  return { ok: true, trigger: { primitive, ref: { kind: 'anchor', anchor } } };
}

/** semantic な PlaceRef → 登録済みの実 PlaceId 群。 */
export function expandPlaceIds(ref: PlaceRef, dict: PlaceDict): PlaceId[] {
  switch (ref.kind) {
    case 'anchor':
      return [ref.anchor];
    case 'place':
      return [ref.placeId];
    case 'label':
      return (dict[ref.key] ?? []).filter(isSimulatable);
  }
}

/** semantic Trigger → 展開後の per-place Trigger 群（監視/バッジ表示に使う）。 */
export function expandTriggers(trigger: Trigger, dict: PlaceDict): Trigger[] {
  if (trigger.primitive === 'time') return [trigger];
  return expandPlaceIds(trigger.ref, dict).map((placeId) => ({
    primitive: trigger.primitive,
    ref: { kind: 'place', placeId },
  }));
}

File: ios/SonotokiKit/Sources/SonotokiKit/Resolver.swift
import Foundation

// Deterministic Resolver — semantic Moment → one of three trigger primitives.
// No AI here. Ported from web/src/domain/resolver.ts, adapted for 1-label-N-places.

public enum ResolveResult: Sendable, Equatable {
    case ok(Trigger)
    /// 独自の呼び方だが辞書に無い → 一度だけ「どこ?」と尋ねる。
    case needsLearning(phrase: String)
    case failed(reason: String)
}

public func resolve(_ c: MomentCandidate, dictionary: PlaceDictionary) -> ResolveResult {
    switch c.kind {
    case .time:
        guard let bucket = c.timeBucket else {
            return .failed(reason: "時間帯が特定できませんでした")
        }
        return .ok(.time(bucket))

    case .homeArrival:
        return .ok(.placeEnter(.anchor(.home)))
    case .leaveHome:
        return .ok(.placeExit(.anchor(.home)))
    case .workArrival:
        return .ok(.placeEnter(.anchor(.work)))

    case .placeArrival, .placeDeparture:
        let make: (PlaceTarget) -> Trigger =
            c.kind == .placeDeparture ? { Trigger.placeExit($0) } : { Trigger.placeEnter($0) }

        if let anchor = c.anchorHint {
            return .ok(make(.anchor(anchor)))
        }
        if let target = c.learnedTarget {
            return .ok(make(target))
        }
        guard let phrase = c.placePhrase ?? c.placeLabel, !phrase.isEmpty else {
            return .failed(reason: "場所が特定できませんでした")
        }
        if let key = dictionary.lookup(phrase) {
            return .ok(make(.label(key)))
        }
        return .needsLearning(phrase: phrase)
    }
}

// MARK: - Expansion（1 semantic ターゲット → 登録済みの複数の実場所 / 複数 Trigger）

/// semantic な `PlaceTarget` → 監視すべき Core Location リージョン識別子（token）群。
/// web の `expandPlaceIds` に対応。CLMonitor の条件セットアップに使う。
public func expandRegionTokens(_ target: PlaceTarget, dictionary: PlaceDictionary) -> [String] {
    switch target {
    case .anchor(let a):
        return ["anchor:\(a.rawValue)"]
    case .region(let token):
        return [token]
    case .label(let key):
        return dictionary.label(forKey: key)?.regionTokens ?? []
    }
}

/// semantic Trigger → 展開後の per-place Trigger 群。web の `expandTriggers` に対応。
public func expandTriggers(_ trigger: Trigger, dictionary: PlaceDictionary) -> [Trigger] {
    switch trigger {
    case .time:
        return [trigger]
    case .placeEnter(let target):
        return expandRegionTokens(target, dictionary: dictionary).map { .placeEnter(.region(token: $0)) }
    case .placeExit(let target):
        return expandRegionTokens(target, dictionary: dictionary).map { .placeExit(.region(token: $0)) }
    }
}
